### DISCUSSION TITLE
Using Tablesample creates SQL syntax error

### DISCUSSION THREAD

**Discussion by sebtrack** [sqlalchemy/sqlalchemy#8214] [Usage Questions] [answered]
Hi,

we run version 1.3.22, so it might be related to that version. I tried very different versions on how to use the tablesample function, the two most promising ones are:

```python
query = session.query(MyModel).with_entities(*[func.distinct(MyModel.col1)])
```

by looking through your tests I've found this way:

```python
sq = query.subquery(key).tablesample(func.bernoulli(0.0001))
end_query = session.query(sq.c.user_agent_fingerprint1)
```

this results in a quite good looking query like this:

```sql
SELECT anon_1.col1
FROM
  (SELECT distinct(MyModel.col1) AS col1
   FROM MyModel) AS anon_1 TABLESAMPLE bernoulli(0.0001)
```

even closer to the test would be using the select statement:

```python
sq = query.subquery(key).tablesample(func.bernoulli(0.0001))
end_select = select([sq.c.user_agent_fingerprint1)])
```

both run into the same sql error:

```
ProgrammingError: (psycopg2.errors.SyntaxError) syntax error at or near "TABLESAMPLE"
LINE 3: FROM MyModel) AS anon_1 TABLESAMPLE bernoulli(0....
                      ^

[SQL: SELECT anon_1.col1 
FROM (SELECT distinct(MyModel.col1) AS col1 
FROM MyModel) AS anon_1 TABLESAMPLE bernoulli(%(bernoulli_1)s)]
[parameters: {'bernoulli_1': 0.0001}]
(Background on this error at: http://sqlalche.me/e/13/f405)
```

using this sql manually yields the same error, it doesn't like the sql alias before the tablesample.

Is there a way to do it correct or is it only possible in 1.4?

EDIT: one thing to note is that `key` (subquery name/alias) is not `anon1` so the anon1 comes from the tablesample function imho.

Thank you!

**Comment by zzzeek** [comment, marked as answer]
OK and...what SQL works ?    need to get a working SQL statement on your target backend, for all I know at the moment is maybe you are on an older PostgreSQL version, or TABLESAMPLE doesnt support subqueries, etc.   I dont actually use this feature.

**Reply by sebtrack** [reply]
Postgres version is: `PostgreSQL 14.3`

A sql statement that works is:

```sql
SELECT
    DISTINCT MyModel.col1
FROM MyModel
TABLESAMPLE BERNOULLI(0.0001);
```

and interestingly though ([your test](https://github.com/sqlalchemy/sqlalchemy/blob/979ea6b21f71605314dc0ac1231dd385eced98c4/test/sql/test_tablesample.py#L58-L65)):

```sql
SELECT alias.id FROM MyModel AS alias TABLESAMPLE bernoulli(:bernoulli_1) REPEATABLE (random())
```

https://github.com/sqlalchemy/sqlalchemy/blob/979ea6b21f71605314dc0ac1231dd385eced98c4/test/sql/test_tablesample.py#L58-L65

**Reply by zzzeek** [reply]
OK so there's just one "alias" there and it's the tablesample.  like this:

```py
const = tablesample(MyModel, func.bernoulli(.0001))
stmt = select(const.c.data.distinct())
```


demo

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import func
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy import tablesample
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String)

e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

s = Session(e)

const = tablesample(A, func.bernoulli(.0001))
stmt = select(const.c.data.distinct())

s.execute(stmt)
```

**Reply by sebtrack** [reply]
Thank you! In this case it would work because the model is "quite close", but could it on a subquery or cte?

**Reply by zzzeek** [reply]
show me working SQL and I will show you how to render it

**Reply by sebtrack** [reply]
Hey, sorry to bother you with it. I came from the perspective of nested subqueries and to have a general working class I was thinking I need this support. I tried to do it in SQL and lo and behold: `TABLESAMPLE clause can only be applied to tables and materialized views`. So, this will never be possible and the Model/Table can be used immediately everytime.

I have a different question though, can I come up with the percentage ('.0001') in a subquery? I can't do it in SQL nor in sqlalchemy, maybe by defining a sql function or declare a variable:

```python
perc = session.query(MyModel).with_entities(*[1000.0/func.count(func.distinct(MyModel.col1))]).scalar()
sq = tablesample(MyModel, func.bernoulli(perc))
session.query(sq.c.col1))
```

`perc` is first retrieved and then the sql statement is build, I wonder if you have an idea how to do it without waiting on the response as the distinct query takes a long time on our table (millions of rows).

**Reply by zzzeek** [reply]
```py
const = tablesample(A, func.bernoulli(
    select(100/func.count(func.distinct(A.data))).scalar_subquery()
))
stmt = select(const.c.data.distinct())

```
